File: Document/MATLAB_LJUDDotNET/Examples/U6/u6_timer_counter.m
%
% Basic U6 example does a PWM output and a counter input features using
% MATLAB, .NET and the UD driver.
%
% support@labjack.com
%

clc %Clear the MATLAB command window
clear %Clear the MATLAB variables

ljasm = NET.addAssembly('LJUDDotNet'); %Make the UD .NET assembly visible in MATLAB
ljudObj = LabJack.LabJackUD.LJUD;

try
    %Read and display the UD version.
    disp(['UD Driver Version = ' num2str(ljudObj.GetDriverVersion())])

    %Open the first found LabJack U6.
    [ljerror, ljhandle] = ljudObj.OpenLabJack(LabJack.LabJackUD.DEVICE.U6, LabJack.LabJackUD.CONNECTION.USB, '0', true, 0);
    
    %First requests to configure the timer and counter.  These will be
    %done with and add/go/get block.
    
    %Set the timer/counter pin offset to 0, which will put the first
    %timer/counter on FIO0.
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_CONFIG, LabJack.LabJackUD.CHANNEL.TIMER_COUNTER_PIN_OFFSET, 0, 0, 0);
    
    %Use the 48 MHz timer clock base with divider (LJ_tc48MHZ_DIV).  Since we are using clock with divisor
    %support, Counter0 is not available.
    LJ_tc48MHZ_DIV = ljudObj.StringToConstant('LJ_tc48MHZ_DIV');
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_CONFIG, LabJack.LabJackUD.CHANNEL.TIMER_CLOCK_BASE, LJ_tc48MHZ_DIV, 0, 0);
    
    %Set the divisor to 48 so the actual timer clock is 1 MHz.
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_CONFIG, LabJack.LabJackUD.CHANNEL.TIMER_CLOCK_DIVISOR, 48, 0, 0);
    
    %Enable 1 timer.  It will use FIO0.
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_CONFIG, LabJack.LabJackUD.CHANNEL.NUMBER_TIMERS_ENABLED, 1, 0, 0);
    
    %Make sure Counter0 is disabled.
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_COUNTER_ENABLE, 0, 0, 0, 0);
    
    %Enable Counter1.  It will use FIO1 since 1 timer is enabled.
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_COUNTER_ENABLE, 1, 1, 0, 0);
    
    %Configure Timer0 as 8-bit PWM (LJ_tmPWM8).  Frequency will be 1M/256 = 3906 Hz.
    LJ_tmPWM8 = ljudObj.StringToConstant('LJ_tmPWM8');
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_TIMER_MODE, 0, LJ_tmPWM8, 0, 0);
    
    %Set the PWM duty cycle to 50%.
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_TIMER_VALUE, 0, 32768, 0, 0);
    
    %Execute the requests.
    ljudObj.GoOne(ljhandle);
    
    %Get all the results just to check for errors.
    [ljerror, ioType, channel, dblValue, dummyInt, dummyDbl] = ljudObj.GetFirstResult(ljhandle, 0, 0, 0, 0, 0);
    
    finished = false;
    while finished == false
        try
            [ljerror, ioType, channel, dblValue, dummyInt, dummyDbl] = ljudObj.GetNextResult(ljhandle, 0, 0, 0, 0, 0);
        catch e
            if(isa(e, 'NET.NetException'))
                eNet = e.ExceptionObject;
                if(isa(eNet, 'LabJack.LabJackUD.LabJackUDException'))
                    %If we get an error, report it.  If the error is NO_MORE_DATA_AVAILABLE we are done
                    if(eNet.LJUDError == LabJack.LabJackUD.LJUDERROR.NO_MORE_DATA_AVAILABLE)
                        finished = true;
                    end
                end
            end
            %Report non NO_MORE_DATA_AVAILABLE error.
            if(finished == false)
                throw(e)
            end
        end
    end
    
    %Wait 1 second.
    pause(1);
    
    %Request a read from the counter.
    [ljerror, dblValue] = ljudObj.eGet(ljhandle, LabJack.LabJackUD.IO.GET_COUNTER, 1, 0, 0);
    
    %This should read roughly 4k counts if FIO0 is shorted to FIO1.
    disp(['Counter 1 = ' num2str(dblValue)]);
    
    %Wait 1 second.
    pause(1);
    
    %Request a read from the counter.
    [ljerror, dblValue] = ljudObj.eGet(ljhandle, LabJack.LabJackUD.IO.GET_COUNTER, 1, 0, 0);
    
    %This should read about 3906 counts more than the previous read.
    disp(['Counter 1 = ' num2str(dblValue)]);
    
    %Disable the timer and counter, and the FIO lines will return to
    %digital I/O. Also setting timer clock base to default 48 Mhz.
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_CONFIG, LabJack.LabJackUD.CHANNEL.NUMBER_TIMERS_ENABLED, 0, 0, 0);
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_COUNTER_ENABLE, 1, 0, 0, 0);
    LJ_tc48MHZ = ljudObj.StringToConstant('LJ_tc48MHZ');
    ljudObj.AddRequest(ljhandle, LabJack.LabJackUD.IO.PUT_CONFIG, LabJack.LabJackUD.CHANNEL.TIMER_CLOCK_BASE, LJ_tc48MHZ, 0, 0);
    ljudObj.GoOne(ljhandle);
    
    %The PWM output sets FIO0 to output, so we do a read here to set
    %it to input.
    ljudObj.eGet(ljhandle, LabJack.LabJackUD.IO.GET_DIGITAL_BIT, 0, 0, 0);
catch e
    showErrorMessage(e)
end

File: DotNet_LJUD/Examples/LJUDDotNet/U6/C#/U6_TimerCounter/U6_TimerCounter.cs
//---------------------------------------------------------------------------
//
//  U6_TimerCounter.cs
// 
//  Basic U6 example does a PWM output and a counter input, using AddGoGet method.
//
//  Please jumper FIO0 to FIO1
//
//  support@labjack.com
//  June 5, 2009
//	Revised March 24, 2015
//----------------------------------------------------------------------
//

using System;
using System.Threading;

// Import the UD .NET wrapper object.  The dll referenced is installed by the
// LabJackUD installer.
using LabJack.LabJackUD; 

namespace TimerCounter
{
	class TimerCounter
	{
		// our U6 variable
		private U6 u6;

		static void Main(string[] args)
		{
			TimerCounter a = new TimerCounter();
			a.performActions();
		}

		// If error occured print a message indicating which one occurred. If the error is a group error (communication/fatal), quit
		public void showErrorMessage(LabJackUDException e)
		{
			Console.Out.WriteLine("Error: " + e.ToString());
			if (e.LJUDError > U6.LJUDERROR.MIN_GROUP_ERROR)
			{
				Console.ReadLine(); // Pause for the user
				Environment.Exit(-1);
			}
		}

		public void performActions()
		{
			LJUD.IO ioType=0;
			LJUD.CHANNEL channel=0;
			double dblValue=0;

			// Variables to satisfy certain method signatures
			int dummyInt = 0;
			double dummyDouble = 0;
			double[] dummyDoubleArray = {0};

			//Open the Labjack with id 2
			try 
			{
				u6 = new U6(LJUD.CONNECTION.USB, "1", true); // Connection through USB

				//First requests to configure the timer and counter.  These will be
				//done with and add/go/get block.

				//Set the timer/counter pin offset to 0, which will put the first
				//timer/counter on FIO0.
				LJUD.AddRequest (u6.ljhandle,  LJUD.IO.PUT_CONFIG, LJUD.CHANNEL.TIMER_COUNTER_PIN_OFFSET, 0, 0, 0);

				//Use the 48 MHz timer clock base with divider.  Since we are using clock with divisor
				//support, Counter0 is not available.
				LJUD.AddRequest(u6.ljhandle, LJUD.IO.PUT_CONFIG, LJUD.CHANNEL.TIMER_CLOCK_BASE, (double)LJUD.TIMERCLOCKS.MHZ48_DIV, 0, 0);

				//Set the divisor to 48 so the actual timer clock is 1 MHz.
				LJUD.AddRequest(u6.ljhandle, LJUD.IO.PUT_CONFIG, LJUD.CHANNEL.TIMER_CLOCK_DIVISOR, 48, 0, 0);

				//Enable 1 timer.  It will use FIO0.
				LJUD.AddRequest(u6.ljhandle, LJUD.IO.PUT_CONFIG, LJUD.CHANNEL.NUMBER_TIMERS_ENABLED, 1, 0, 0);

				//Make sure Counter0 is disabled.
				LJUD.AddRequest(u6.ljhandle, LJUD.IO.PUT_COUNTER_ENABLE, 0, 0, 0, 0);

				//Enable Counter1.  It will use FIO1 since 1 timer is enabled.
				LJUD.AddRequest(u6.ljhandle, LJUD.IO.PUT_COUNTER_ENABLE, 1, 1, 0, 0);

				//Configure Timer0 as 8-bit PWM.  Frequency will be 1M/256 = 3906 Hz.
				LJUD.AddRequest(u6.ljhandle, LJUD.IO.PUT_TIMER_MODE, 0, (double)LJUD.TIMERMODE.PWM8, 0, 0);

				//Set the PWM duty cycle to 50%.
				LJUD.AddRequest(u6.ljhandle, LJUD.IO.PUT_TIMER_VALUE, 0, 32768, 0, 0);

				//Execute the requests.
				LJUD.GoOne (u6.ljhandle);
			}
			catch (LabJackUDException e) 
			{
				showErrorMessage(e);
			}

			//Get all the results just to check for errors.
			try {LJUD.GetFirstResult(u6.ljhandle, ref ioType, ref channel, ref dblValue, ref dummyInt, ref dummyDouble);}
			catch (LabJackUDException e) {showErrorMessage(e);}
			bool finished = false;
			while(!finished)
			{
				try{LJUD.GetNextResult(u6.ljhandle, ref ioType, ref channel, ref dblValue, ref dummyInt, ref dummyDouble);}
				catch (LabJackUDException e) 
				{
					if (e.LJUDError == LJUD.LJUDERROR.NO_MORE_DATA_AVAILABLE)
						finished = true;
					else
						showErrorMessage(e);
				}
			}

			try
			{
				//Wait 1 second.
				Thread.Sleep(1000);

				//Request a read from the counter.
				LJUD.eGet(u6.ljhandle, LJUD.IO.GET_COUNTER, (LJUD.CHANNEL)1, ref dblValue, dummyDoubleArray);
  
				//This should read roughly 4k counts if FIO0 is shorted to FIO1.
				Console.Out.WriteLine("Counter = {0:0.0}\n",dblValue);

				//Wait 1 second.
				Thread.Sleep(1000);

				//Request a read from the counter.
				LJUD.eGet(u6.ljhandle, LJUD.IO.GET_COUNTER, (LJUD.CHANNEL)1, ref dblValue, dummyDoubleArray);

				//This should read about 3906 counts more than the previous read.
				Console.Out.WriteLine("Counter = {0:0.0}\n",dblValue);

				//Reset all pin assignments to factory default condition.
				LJUD.ePut (u6.ljhandle, LJUD.IO.PIN_CONFIGURATION_RESET, 0, 0, 0);

				//The PWM output sets FIO0 to output, so we do a read here to set
				//it to input.
				LJUD.eGet (u6.ljhandle, LJUD.IO.GET_DIGITAL_BIT, 0, ref dblValue, 0);
			}
			catch (LabJackUDException e) 
			{
				if (e.LJUDError == LJUD.LJUDERROR.NO_MORE_DATA_AVAILABLE)
					finished = true;
				else
					showErrorMessage(e);
			}

			Console.ReadLine(); // Pause for user	
		}
	}
}
